File: rust/crates/crispy-core/src/traits/profile_repository.rs
use std::collections::HashMap;

use crate::database::DbError;
use crate::models::UserProfile;

/// Persistence contract for user profiles, source access, and channel order.
pub trait ProfileRepository {
    fn save_profile(&self, profile: &UserProfile) -> Result<(), DbError>;
    fn delete_profile(&self, id: &str) -> Result<(), DbError>;
    fn load_profiles(&self) -> Result<Vec<UserProfile>, DbError>;
    fn grant_source_access(
        &self,
        profile_id: &str,
        source_id: &str,
    ) -> Result<(), DbError>;
    fn revoke_source_access(
        &self,
        profile_id: &str,
        source_id: &str,
    ) -> Result<(), DbError>;
    fn get_source_access(&self, profile_id: &str) -> Result<Vec<String>, DbError>;
    fn set_source_access(
        &self,
        profile_id: &str,
        source_ids: &[String],
    ) -> Result<(), DbError>;
    fn save_channel_order(
        &self,
        profile_id: &str,
        group_name: &str,
        channel_ids: &[String],
    ) -> Result<(), DbError>;
    fn load_channel_order(
        &self,
        profile_id: &str,
        group_name: &str,
    ) -> Result<Option<HashMap<String, i32>>, DbError>;
    fn reset_channel_order(
        &self,
        profile_id: &str,
        group_name: &str,
    ) -> Result<(), DbError>;
}

File: app/flutter/lib/features/profiles/domain/repositories/profile_repository.dart
import '../entities/user_profile.dart';

/// Repository contract for core user profile and profile-scoped
/// preference operations.
///
/// Implemented by the infrastructure layer backed by the Rust
/// crispy-core engine via CacheService.
///
/// Favorites operations are defined in [ProfileFavoritesRepository].
abstract interface class ProfileRepository {
  // ── Profiles ───────────────────────────────────────

  /// Save a user profile (upsert).
  Future<void> saveProfile(UserProfile profile);

  /// Delete a profile and its associated data by [id].
  Future<void> deleteProfile(String id);

  /// Load all user profiles.
  Future<List<UserProfile>> loadProfiles();

  // ── Source Access ──────────────────────────────────

  /// Grant a profile access to a source.
  Future<void> grantSourceAccess(String profileId, String sourceId);

  /// Revoke a profile's access to a source.
  Future<void> revokeSourceAccess(String profileId, String sourceId);

  /// Get all source IDs accessible to [profileId].
  Future<List<String>> getSourceAccess(String profileId);

  /// Set source access for [profileId], replacing any existing list.
  Future<void> setSourceAccess(String profileId, List<String> sourceIds);

  /// Get all profile IDs with access to [sourceId].
  Future<List<String>> getProfilesWithSourceAccess(String sourceId);

  // ── Channel Custom Order ───────────────────────────

  /// Save a custom channel order for a specific group and profile.
  Future<void> saveChannelOrder(
    String profileId,
    String groupName,
    List<String> channelIds,
  );

  /// Load the custom channel order for a specific group and profile.
  ///
  /// Returns null when no custom order has been saved.
  Future<Map<String, int>?> loadChannelOrder(
    String profileId,
    String groupName,
  );

  /// Reset the custom channel order for a specific group and profile.
  Future<void> resetChannelOrder(String profileId, String groupName);
}
